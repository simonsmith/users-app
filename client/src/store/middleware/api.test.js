import axios from 'axios';
import apiMiddleware, {makeApiRequest} from './api';
import configureMockStore from 'redux-mock-store';

jest.mock('axios');

const mockStore = configureMockStore([apiMiddleware]);

afterEach(() => {
  axios.mockClear();
});

test("ignores actions that do not match 'API_REQUEST'", () => {
  const store = mockStore({});
  store.dispatch({type: 'FOO'});
  expect(store.getActions()).toEqual([{type: 'FOO'}]);
});

test('handles http success and dispatches actions with data as the payload', async () => {
  axios.mockResolvedValue({
    data: 'some data',
  });
  const store = mockStore({});

  await store.dispatch({
    API_REQUEST: {
      type: 'TEST',
      meta: {
        url: 'http://foo.com',
        method: 'get',
      },
    },
  });

  expect(axios).toHaveBeenCalledWith({
    url: 'http://foo.com',
    method: 'get',
  });
  expect(store.getActions()).toEqual([
    {type: 'TEST_REQUEST'},
    {type: 'TEST_SUCCESS', payload: 'some data'},
  ]);
});

test('handles http failure and dispatches actions with error as the payload', async () => {
  axios.mockRejectedValue(new Error('api down'));
  const store = mockStore({});
  await store.dispatch({
    API_REQUEST: {
      type: 'TEST',
      meta: {
        url: 'http://foo.com',
      },
    },
  });
  expect(store.getActions()).toEqual([
    {type: 'TEST_REQUEST'},
    {type: 'TEST_FAILURE', error: true, payload: new Error('api down')},
  ]);
});

test('uses optional transformer on data before dispatching', async () => {
  axios.mockResolvedValue({
    data: 'some data',
  });
  const store = mockStore({});
  const transformerMock = jest.fn().mockReturnValue('transformed data');

  await store.dispatch({
    API_REQUEST: {
      type: 'TEST',
      meta: {
        url: 'http://foo.com',
        transformer: transformerMock,
      },
    },
  });

  expect(transformerMock).toHaveBeenCalledWith('some data');
  expect(store.getActions()).toEqual([
    {type: 'TEST_REQUEST'},
    {type: 'TEST_SUCCESS', payload: 'transformed data'},
  ]);
});

test('makeApiRequest returns a correctly formatted API_REQUEST action object', () => {
  expect(
    makeApiRequest({url: 'foo', type: 'FOO', transformer: jest.fn()})
  ).toMatchSnapshot();
});
