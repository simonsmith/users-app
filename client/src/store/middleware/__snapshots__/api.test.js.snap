// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeApiRequest returns a correctly formatted API_REQUEST action object 1`] = `
Object {
  "API_REQUEST": Object {
    "meta": Object {
      "method": "get",
      "transformer": [MockFunction],
      "url": "foo",
    },
    "type": "FOO",
  },
}
`;
